[IM] Index Templates UI list view

alisonelizabeth opened the pull request:
This PR is the first phase of adding support for `Index Templates` within the `Index Management` app.

Changes include:
- New tabbed content under Index Management - `Indices` & `Index Templates`
- List view of a user's index templates with delete action + filter by ILM policy
- API integration tests for fetching, deleting and creating index templates
- Component integration tests for index templates list view
- Telemetry for index templates list view

## Release notes
This feature adds the list of Index Templates as a tab to the Index Management app. Users can also delete templates from this view.

*Note*: This is the first phase, additional functionality for Index Templates will be added for 7.4 in separate PRs.

## Testing
There are several system index templates by default. To create your own index template, go to console and follow the [ES API docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-templates.html).

For example:
```
PUT _template/template_1
{
  "index_patterns": ["te*", "bar*"],
  "settings": {
    "number_of_shards": 1
  },
  "mappings": {
    "_source": {
      "enabled": false
    },
    "properties": {
      "host_name": {
        "type": "keyword"
      },
      "created_at": {
        "type": "date",
        "format": "EEE MMM dd HH:mm:ss Z yyyy"
      }
    }
  }
}
```

## Screenshots
_Before_
<img width="1233" alt="Screen Shot 2019-07-08 at 1 48 12 PM" src="https://user-images.githubusercontent.com/5226211/60831218-123aa180-a187-11e9-9d89-854b31b4f355.png">

_After_
**Indices (functionality unchanged, but now under `Indices` tab)**

<img width="1249" alt="Screen Shot 2019-07-12 at 2 50 15 PM" src="https://user-images.githubusercontent.com/5226211/61151632-6493ff00-a4b4-11e9-84ed-39d511e44ba2.png">


**Index Templates (new)**

List view w/o system templates:
<img width="1230" alt="Screen Shot 2019-07-10 at 2 57 46 PM" src="https://user-images.githubusercontent.com/5226211/60996808-6084a780-a323-11e9-9ffb-fcc2b9203385.png">

List view with system templates toggled:
<img width="1232" alt="Screen Shot 2019-07-12 at 2 47 20 PM" src="https://user-images.githubusercontent.com/5226211/61151470-023afe80-a4b4-11e9-92d2-92b188b801fd.png">

Filtered list view by ILM policy:
<img width="1227" alt="Screen Shot 2019-07-10 at 2 58 04 PM" src="https://user-images.githubusercontent.com/5226211/60996838-72664a80-a323-11e9-9286-354d28166b9a.png">

Delete single template:
<img width="1248" alt="Screen Shot 2019-07-10 at 2 58 38 PM" src="https://user-images.githubusercontent.com/5226211/60996846-785c2b80-a323-11e9-8d27-2a25a088f42e.png">

Delete multiple templates:
<img width="1256" alt="Screen Shot 2019-07-10 at 2 58 28 PM" src="https://user-images.githubusercontent.com/5226211/60996871-7f833980-a323-11e9-86f0-0ad667d7c987.png">

Delete system template:
<img width="1239" alt="Screen Shot 2019-07-12 at 2 48 54 PM" src="https://user-images.githubusercontent.com/5226211/61151551-331b3380-a4b4-11e9-954e-2156809f22c8.png">

elasticmachine commented:
Pinging @elastic/es-ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4733/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5130/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5426/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5619/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6134/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6143/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6150/)

alisonelizabeth commented:
Retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6162/)

Conky5 commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6202/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6417/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6450/)

alisonelizabeth commented:
FYI @yaronp68 this is ready for review

alisonelizabeth commented:
@Bamieh do you have capacity to review `i18n` on this PR?

cjcenizal commented:
Great work Alison! I took a look at the UI locally and it's looking terrific. I had a couple questions. Is there a reason to not align the switches on the right using a flex group? I think this layout is a bit better because these switches are meant to be used less often than other parts of the UI, and putting them on the right side moves them a bit out of the way.

![image](https://user-images.githubusercontent.com/1238659/60987057-ae82b680-a2f5-11e9-81c2-a05ed4dad3ba.png)

My other question was about deleting system index templates. We let the user delete system indices, so should we also let them delete system index templates with the same "thoroughly warn the user" UX?

alisonelizabeth commented:
@cjcenizal thanks for the feedback!

> Is there a reason to not align the switches on the right using a flex group?

thanks for calling that out - there was not a particular reason. i agree i think that looks better. i'll make the change.

> My other question was about deleting system index templates. We let the user delete system indices, so should we also let them delete system index templates with the same "thoroughly warn the user" UX?

I had initially thought this as well, however I spoke with @yaronp68 regarding it and he felt the action should be out of reach for users as deleting a template would mean new system indices cannot be created. Thoughts?

cjcenizal commented:
> he felt the action should be out of reach for users as deleting a template would mean new system indices cannot be created.

Thanks, I understand. I agree this is not something we want users to do, but I think it's also something users should be able to do if they really know what they're doing. So my suggestion would be to treat this the same way as system indices and force users to confirm that they want to do something terrible. I think it's OK to punt on this for now if you'd like to discuss with Yaron some more, though, so I don't consider this a blocker.

If we do decide to implement my suggestion, maybe we could address Yaron's concern by adding a feature control that lets admins disable the ability to delete system indices for certain users.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6670/)

alisonelizabeth commented:
@cjcenizal Thanks, I'll discuss more with Yaron.

Also, fixed the alignment with the switches and updated the screenshots.

alisonelizabeth commented on x-pack/legacy/plugins/index_management/public/services/use_request.ts:
```diff
@@ -0,0 +1,131 @@
+/*
```
I'll delete this file and adopt the shared plugin once https://github.com/elastic/kibana/pull/40777 is merged.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6715/)

yaronp68 commented:
@cjcenizal @alisonelizabeth I'm fine with a confirmation approach for deleting system templates but worth checking with support later. I'll ask

alisonelizabeth commented:
updated PR to allow deletion of system templates with confirmation (see latest screenshots in first comment)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7205/)

sebelga commented on x-pack/legacy/plugins/index_management/public/components/delete_templates_modal.tsx:
```diff
@@ -0,0 +1,169 @@
+    templatesToDelete.find(templateName => templateName.startsWith('.'))
+  );
+
+  const handleDeleteTemplates = async () => {
```
nit: the `async` does not seem to be needed here

sebelga commented on x-pack/legacy/plugins/index_management/public/components/delete_templates_modal.tsx:
```diff
@@ -0,0 +1,169 @@
+  );
+
+  const handleDeleteTemplates = async () => {
+    deleteTemplates(templatesToDelete).then(({ data: { templatesDeleted, errors }, error }) => {
```
I haven't looked at the API yet, but it is a bit confusing to have both "errors" and "error". It does seem though that is is a pattern we have used in another app though (now I am the reader and not the writer of the code 😊 ). If `error` (singular) is the Http request that failed, could we rename it "apiError"?

sebelga commented on x-pack/legacy/plugins/index_management/public/components/delete_templates_modal.tsx:
```diff
@@ -0,0 +1,169 @@
+
+  const [isDeleteConfirmed, setIsDeleteConfirmed] = useState<boolean>(false);
+
+  if (!numTemplatesToDelete) {
```
This would be better on L23 to exit early.

sebelga commented on x-pack/legacy/plugins/index_management/public/sections/home/templates_list/templates_table/templates_table.tsx:
```diff
@@ -0,0 +1,247 @@
+const Checkmark = ({ tableCellData }: { tableCellData: object }) => {
+  const isChecked = Object.entries(tableCellData).length > 0;
+
+  if (isChecked) {
```
nit: In these cases, I usually personally prefer the shorter ternary version

```js
return isChecked ? <EuiIcon type="check" /> : null;
```

sebelga commented on x-pack/legacy/plugins/index_management/server/routes/api/templates/register_delete_route.ts:
```diff
@@ -0,0 +1,41 @@
+      } catch (e) {
+        return response.errors.push({
+          name,
+          error: wrapEsError(e),
```
Now I see it 😊

Bamieh commented on x-pack/legacy/plugins/index_management/public/components/delete_templates_modal.tsx:
```diff
@@ -0,0 +1,169 @@
+          'xpack.idxMgmt.deleteTemplatesModal.successNotificationMessageText',
+          {
+            defaultMessage:
+              'Deleted {numSuccesses, number} {numSuccesses, plural, one {template} other {templates}}',
```
this is OK, but i'd use this:

```suggestion
              'Deleted {numSuccesses, plural, one {# template} other {# templates}}',
```

the `#` will return the variable value.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8103/)

alisonelizabeth commented:
@sebelga @ahmad thanks for your reviews! I addressed the feedback, so going to go ahead and merge.

@sebelga I agree the ILM button is a little weird, but that is a EUI decision.
